hudi-client/src/main/java/org/apache/hudi/io/HoodieCommitArchiveLog.java

```diff
@@ -169,7 +180,14 @@ public boolean archiveIfRequired(final JavaSparkContext jsc) throws IOException
       }).limit(commitTimeline.countInstants() - minCommitsToKeep));
     }
 
-    return instants;
+    // For archiving and cleaning instants, we need to include intermediate state files if they exist
+    HoodieActiveTimeline rawActiveTimeline = new HoodieActiveTimeline(metaClient, false);
+    Map<Pair<String, String>, List<HoodieInstant>> groupByTsAction = rawActiveTimeline.getInstants()
+        .collect(Collectors.groupingBy(x -> Pair.of(x.getTimestamp(),
+            x.getAction().equals(HoodieTimeline.COMPACTION_ACTION) ? HoodieTimeline.COMMIT_ACTION : x.getAction())));
+
+    return instants.flatMap(hoodieInstant ->
```

Comment:
Should we filter only the necessary instants to group by  ?

rawActiveTimeline.getInstants().filter(i -> instants.contains(i)), this way the map will be smaller..

Reply:
What I'm saying is group by only the instants chosen to archive, all types of actions is fine..